### PR TITLE
Course 7

### PR DESCRIPTION
**Pull request by AnnaMartis**

### CODE REVIEW

File: src/entities/MovieForm/ui/MovieForm.jsx [hunk 1]
```diff
@@ -1,106 +1,165 @@
...
+import { useOutletContext } from "react-router";
 import "./MovieForm.css";
 import { genresList } from "../../../entities/SelectTabs/selectTabsMock";
 import { getMovie } from "../lib/getMovie";
+import { useForm, Controller } from "react-hook-form";
 
-export const MovieForm = ({ initialMovie, onFormSubmit }) => {
-  const [movie, setMovie] = useState(getMovie(initialMovie));
+export const MovieForm = ({ initialMovie }) => {
+  const { movie } = useOutletContext() ?? {initialMovie};
```
Comment (dmitriisurdin) [inline src/entities/MovieForm/ui/MovieForm.jsx:10]:
```suggestion
  const { movie =  initialMovie} = useOutletContext();
```

File: src/entities/MovieForm/ui/MovieForm.jsx [hunk 2]
```diff
@@ -1,106 +1,165 @@
...
-    }));
+  const { handleSubmit, control, reset } = useForm({
+    defaultValues: getMovie(movie),
+  });
+
+  const onFormSubmit = (data) => {
+    if (movie) {
+      console.log("send request to edit a movie");
+    } else {
+      console.log("send a request to add a new movie");
```
Comment (dmitriisurdin) [inline src/entities/MovieForm/ui/MovieForm.jsx]: do we need it?

File: src/entities/MovieForm/ui/MovieForm.jsx [hunk 3]
```diff
@@ -1,106 +1,165 @@
...
+    }
   };
 
   return (
-    <form className="movie-form" action="#" onSubmit={onFormSubmit}>
+    <form
+      className="movie-form"
+      action="#"
```
Comment (dmitriisurdin) [inline src/entities/MovieForm/ui/MovieForm.jsx]: do we need this attribute here?

File: src/entities/MovieForm/ui/MovieForm.jsx [hunk 4]
```diff
@@ -1,106 +1,165 @@
...
+                  {...field}
+                />
+              );
+            }}
           />
         </div>
       </div>
       <div className="movie-footer">
-        <button className="movie-reset-button">RESET</button>
+        <input className="movie-reset-button" onClick={()=>reset(getMovie(movie))} type="button" value={"Reset"}/>
```
Comment (dmitriisurdin) [inline src/entities/MovieForm/ui/MovieForm.jsx]:
```suggestion
        <input className="movie-reset-button" onClick={()=>reset(getMovie(movie))} type="button" value="Reset"/>
```

Comment (dmitriisurdin) [reply src/entities/MovieForm/ui/MovieForm.jsx]: also why wouldn't we just use tag `button` here?

File: src/entities/MovieTitle/MovieTitle.jsx [hunk 5]
```diff
@@ -3,6 +3,7 @@ import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
 import { faEllipsisVertical } from "@fortawesome/free-solid-svg-icons";
 import { useState } from "react";
 import cx from "classnames";
+import {Link} from 'react-router-dom';
```
Comment (dmitriisurdin) [inline src/entities/MovieTitle/MovieTitle.jsx]:
```suggestion
import { Link } from 'react-router-dom';
```

File: src/entities/SearchFormWrapper/SearchFormWrapper.jsx [hunk 6]
```diff
@@ -1,14 +1,25 @@
...
+  const handleAddMovieCLick = () => {
+    navigate("/new");
+  };
 
   return (
     <>
+      <Outlet />
       <div className="add-movie-part">
-        <CustomButton className="outlined-button" label="+ADD MOVIE" />
+        <CustomButton
```
Comment (dmitriisurdin) [inline src/entities/SearchFormWrapper/SearchFormWrapper.jsx]: looks like it should be a `Link` tag from react-router

File: src/entities/MovieForm/ui/MovieForm.jsx (hunk 1 above)
Comment (AnnaMartis) [reply src/entities/MovieForm/ui/MovieForm.jsx:10]: done

File: src/entities/MovieForm/ui/MovieForm.jsx (hunk 2 above)
Comment (AnnaMartis) [reply src/entities/MovieForm/ui/MovieForm.jsx]: I have added the requests part.

Comment (AnnaMartis) [reply src/entities/MovieForm/ui/MovieForm.jsx]: done

File: src/entities/MovieForm/ui/MovieForm.jsx (hunk 3 above)
Comment (AnnaMartis) [reply src/entities/MovieForm/ui/MovieForm.jsx]: no, done

File: src/entities/MovieForm/ui/MovieForm.jsx (hunk 4 above)
Comment (AnnaMartis) [reply src/entities/MovieForm/ui/MovieForm.jsx]: idk, for a moment it submitted on pressing this button, but now it works

Comment (AnnaMartis) [reply src/entities/MovieForm/ui/MovieForm.jsx]: done

File: src/entities/MovieTitle/MovieTitle.jsx (hunk 5 above)
Comment (AnnaMartis) [reply src/entities/MovieTitle/MovieTitle.jsx]: done

File: src/entities/MovieForm/ui/MovieForm.jsx (hunk 1 above)
Comment (AnnaMartis) [reply src/entities/MovieForm/ui/MovieForm.jsx:10]: it throws an error in this case

File: src/entities/SearchFormWrapper/SearchFormWrapper.jsx (hunk 6 above)
Comment (AnnaMartis) [reply src/entities/SearchFormWrapper/SearchFormWrapper.jsx]: done

File: src/app/App.js [hunk 7]
```diff
@@ -15,12 +13,33 @@ const router = createBrowserRouter([
...
       },
       {
         path: "/:movieId",
         loader: movieLoader,
         element: <MovieDetails />,
+        children: [
+          {
+            path: "/:movieId/edit",
+            element: (
+              <Dialog title={"Edit Movie"}>
```
Comment (dmitriisurdin) [inline src/app/App.js]:
```suggestion
              <Dialog title="Edit Movie">
```

File: src/app/App.js [hunk 8]
```diff
@@ -15,12 +13,33 @@ const router = createBrowserRouter([
     children: [
       {
         path: "/",
-        element: <SearchFormWrapper/>
+        element: <SearchFormWrapper />,
+        children: [
+          {
+            path: "/new",
+            element: (
+              <Dialog title={"Add Movie"}>
```
Comment (dmitriisurdin) [inline src/app/App.js]:
```suggestion
              <Dialog title="Add Movie">
```

File: src/app/App.js [hunk 9]
```diff
@@ -15,12 +13,33 @@ const router = createBrowserRouter([
     children: [
       {
         path: "/",
-        element: <SearchFormWrapper/>
+        element: <SearchFormWrapper />,
+        children: [
+          {
+            path: "/new",
```
Comment (dmitriisurdin) [inline src/app/App.js]: we can move all `path` into a separate file

File: src/entities/MovieForm/ui/MovieForm.jsx [hunk 10]
```diff
@@ -1,106 +1,183 @@
...
-  const handleFieldChange = (event) => {
-    setMovie((prevMovie) => ({
-      ...prevMovie,
-      [event.target.name]: event.target.value,
-    }));
+  const onFormSubmit = async (data) => {
+    try {
+      let requestOptions;
+      if (movie) {
+        requestOptions = {
```
Comment (dmitriisurdin) [inline src/entities/MovieForm/ui/MovieForm.jsx]: code duplication (DRY), only method is different, can be moved to a template

File: src/entities/MovieForm/ui/MovieForm.jsx [hunk 11]
```diff
@@ -1,106 +1,183 @@
...
+      } else {
+        requestOptions = {
+          method: "PUT",
+          headers: {
+            "Content-Type": "application/json",
+          },
+          body: JSON.stringify(data),
+        };
+      }
+      await fetch("http://localhost:4000", requestOptions);
```
Comment (dmitriisurdin) [inline src/entities/MovieForm/ui/MovieForm.jsx]: can be moved to a variable `http://localhost:4000`

File: src/entities/MovieForm/lib/getMovie.js [hunk 12]
```diff
@@ -1,11 +1,11 @@
...
+        title: movie?.title ?? movie?.movieName ?? "",
+        movieUrl: movie?.movieUrl ?? movie?.imgUrl ?? "",
+        genre: movie?.genre ?? "ALL",
+        overview: movie?.overview ?? movie?.description ?? "",
+        releaseDate: movie?.releaseDate ?? movie?.releaseYear ?? "",
         rating: movie?.rating ?? "",
-        runtime: movie?.runtime ?? "",
+        runtime: movie?.runtime ?? movie?.duration ?? "",
       })
 }
```
Comment (dmitriisurdin) [inline src/entities/MovieForm/lib/getMovie.js]:
```suggestion
}

```

File: src/app/App.js (hunk 7 above)
Comment (AnnaMartis) [reply src/app/App.js]: done

File: src/app/App.js (hunk 8 above)
Comment (AnnaMartis) [reply src/app/App.js]: done

File: src/app/App.js (hunk 9 above)
Comment (AnnaMartis) [reply src/app/App.js]: i hope i have done it right

File: src/entities/MovieForm/ui/MovieForm.jsx (hunk 10 above)
Comment (AnnaMartis) [reply src/entities/MovieForm/ui/MovieForm.jsx]: done

File: src/entities/MovieForm/ui/MovieForm.jsx (hunk 11 above)
Comment (AnnaMartis) [reply src/entities/MovieForm/ui/MovieForm.jsx]: done

File: src/entities/MovieForm/lib/getMovie.js (hunk 12 above)
Comment (AnnaMartis) [reply src/entities/MovieForm/lib/getMovie.js]: done

File: src/app/App.js [hunk 13]
```diff
@@ -1,27 +1,15 @@
...
-      {
-        path: "/",
-        element: <SearchFormWrapper/>
-      },
-      {
-        path: "/:movieId",
-        loader: movieLoader,
-        element: <MovieDetails />,
-      },
+      moviecrudPaths
```
Comment (dmitriisurdin) [inline src/app/App.js]: I didn't mean to separate the whole object, I meant only paths, like:
```
export const paths = {
home: '/',
movieDetails: '/:movieId', 
...
}
```
we can store it in `shared/routing/paths.js`

File: src/entities/Dialog/Dialog.jsx [hunk 14]
```diff
@@ -1,15 +1,15 @@
 import { Portal } from "react-portal";
+import {Link} from "react-router-dom";
```
Comment (dmitriisurdin) [inline src/entities/Dialog/Dialog.jsx]:
```suggestion
import { Link } from "react-router-dom";
```

File: src/pages/MovieListPage/MoviesListPage.jsx [hunk 15]
```diff
@@ -67,7 +68,7 @@ export const MovieListPage = () => {
           search: searchQuery,
           filter: [selectedGenre],
         });
-        const response = await fetch(`http://localhost:4000?${queryParams}`);
+        const response = await fetch(MAIN_API +`?${queryParams}`);
```
Comment (dmitriisurdin) [inline src/pages/MovieListPage/MoviesListPage.jsx]:
```suggestion
        const response = await fetch(`${MAIN_API}?${queryParams}`);
```

File: src/entities/MovieForm/ui/MovieForm.jsx [hunk 16]
```diff
@@ -1,106 +1,176 @@
...
+  });
 
-  const handleFieldChange = (event) => {
-    setMovie((prevMovie) => ({
-      ...prevMovie,
-      [event.target.name]: event.target.value,
-    }));
+  const onFormSubmit = async (data) => {
+    let requestOptions = {
+      method: "POST",
```
Comment (dmitriisurdin) [inline src/entities/MovieForm/ui/MovieForm.jsx]:
```suggestion
      method: movie ? "POST" : "PUT",
```

File: src/entities/MovieForm/ui/MovieForm.jsx [hunk 17]
```diff
@@ -1,106 +1,176 @@
...
+      headers: {
```
Comment (dmitriisurdin) [inline src/entities/MovieForm/ui/MovieForm.jsx:19]: in general, it should be managed in one place not every time for each request, but for now it's okay, let's keep

File: src/app/App.js (hunk 9 above)
Comment (dmitriisurdin) [reply src/app/App.js]: not really, I've left another comment

File: src/app/App.js (hunk 13 above)
Comment (AnnaMartis) [reply src/app/App.js]: done

File: src/entities/Dialog/Dialog.jsx (hunk 14 above)
Comment (AnnaMartis) [reply src/entities/Dialog/Dialog.jsx]: done

File: src/pages/MovieListPage/MoviesListPage.jsx (hunk 15 above)
Comment (AnnaMartis) [reply src/pages/MovieListPage/MoviesListPage.jsx]: done

File: src/entities/MovieForm/ui/MovieForm.jsx (hunk 16 above)
Comment (AnnaMartis) [reply src/entities/MovieForm/ui/MovieForm.jsx]: done

File: src/entities/MovieForm/ui/MovieForm.jsx (hunk 17 above)
Comment (AnnaMartis) [reply src/entities/MovieForm/ui/MovieForm.jsx:19]: okaay

**Review by dmitriisurdin — APPROVED** [review]

**Pull request merged**